python/pyiceberg/manifest.py

```diff
@@ -79,88 +95,107 @@ def __repr__(self) -> str:
-    ),
-    NestedField(
-        field_id=102,
-        name="partition",
-        field_type=StructType(),
-        required=True,
-        doc="Partition data tuple, schema based on the partition spec",
-    ),
-    NestedField(field_id=103, name="record_count", field_type=LongType(), required=True, doc="Number of records in the file"),
-    NestedField(field_id=104, name="file_size_in_bytes", field_type=LongType(), required=True, doc="Total file size in bytes"),
-    NestedField(
-        field_id=108,
-        name="column_sizes",
-        field_type=MapType(key_id=117, key_type=IntegerType(), value_id=118, value_type=LongType()),
-        required=False,
-        doc="Map of column id to total size on disk",
-    ),
-    NestedField(
-        field_id=109,
-        name="value_counts",
-        field_type=MapType(key_id=119, key_type=IntegerType(), value_id=120, value_type=LongType()),
-        required=False,
-        doc="Map of column id to total count, including null and NaN",
-    ),
-    NestedField(
-        field_id=110,
-        name="null_value_counts",
-        field_type=MapType(key_id=121, key_type=IntegerType(), value_id=122, value_type=LongType()),
-        required=False,
-        doc="Map of column id to null value count",
-    ),
-    NestedField(
-        field_id=137,
-        name="nan_value_counts",
-        field_type=MapType(key_id=138, key_type=IntegerType(), value_id=139, value_type=LongType()),
-        required=False,
-        doc="Map of column id to number of NaN values in the column",
-    ),
-    NestedField(
-        field_id=125,
-        name="lower_bounds",
-        field_type=MapType(key_id=126, key_type=IntegerType(), value_id=127, value_type=BinaryType()),
-        required=False,
-        doc="Map of column id to lower bound",
-    ),
-    NestedField(
-        field_id=128,
-        name="upper_bounds",
-        field_type=MapType(key_id=129, key_type=IntegerType(), value_id=130, value_type=BinaryType()),
-        required=False,
-        doc="Map of column id to upper bound",
-    ),
-    NestedField(field_id=131, name="key_metadata", field_type=BinaryType(), required=False, doc="Encryption key metadata blob"),
-    NestedField(
-        field_id=132,
-        name="split_offsets",
-        field_type=ListType(element_id=133, element_type=LongType(), element_required=True),
-        required=False,
-        doc="Splittable offsets",
-    ),
-    NestedField(
-        field_id=135,
-        name="equality_ids",
-        field_type=ListType(element_id=136, element_type=LongType(), element_required=True),
-        required=False,
-        doc="Equality comparison field IDs",
-    ),
-    NestedField(field_id=140, name="sort_order_id", field_type=IntegerType(), required=False, doc="Sort order ID"),
-    NestedField(field_id=141, name="spec_id", field_type=IntegerType(), required=False, doc="Partition spec ID"),
-)
+def data_file_type(partition_type: StructType) -> StructType:
```

Comment:
Migrated review:
> I'm not a fan of this one, but I see why it is necessary. For reading, we can override certain field IDs:
> 
> https://github.com/apache/iceberg/blob/e389e4d139624a49729379acd330dd9c96187b04/python/pyiceberg/manifest.py#L319-L335
> 
> We could do the same when writing. We can override field-id `102` when constructing the writer. WDYT?

Reply:
Thanks for the suggestion. I gave it some thought and realized that the Reader only needs the full schema when constructing the reader, thus we could override some fields in the resolver. 

The Writer, on the other hand, needs to have the whole schema in the header. This just means we'll need a handy function to apply the overrides and get our final schema. So, I think it better to stick with this approach for now (as I couldn't think of a better one).

Reply:
The thing is that we're creating many new objects which is not really needed.

Reply:
What do you think of the following approach:

```python
DATA_FILE_TYPE = StructType(
    NestedField(
        field_id=134,
        name="content",
        field_type=IntegerType(),
        required=False,
        doc="Contents of the file: 0=data, 1=position deletes, 2=equality deletes",
        initial_default=DataFileContent.DATA,
    ),
...
    NestedField(
        field_id=102,
        name="partition",
        field_type=StructType(),
        required=True,
        doc="Partition data tuple, schema based on the partition spec",
    ),
...
)

def datafile_with_partition(partition_type: StructType) -> StructType()
    return StructType(
        *[NestedField(
            field_id=102,
            name="partition",
            field_type=partition_type,
            required=True,
            doc="Partition data tuple, schema based on the partition spec",
        ) if field.field_id = 102 else field for field in DATA_FILE_TYPE.fields]
    )
```

This keeps the old `DATA_FILE_TYPE` in place, and allows the ability to create new ones with any partition spec, and we don't recreate all the fields all the time.

Reply:
Thanks for the explanation! I updated the code as suggested and add one more argument `format_version` to `data_file_with_partition` so that we only need one function.